### PR TITLE
Adding UTs to send_msg()

### PR DESCRIPTION
**Pull request by jftuduri**
|Related issue|
|---|
|#13900|

## Description
This PR adds tests in `test_sendmsg.c` to improve coverage to `send_msg` function from `remoted/sendmsg.c`.

### New coverage
new coverage_cropped

### Output
```
$ ./test_sendmsg 
[==========] tests: Running 10 test(s).
[ RUN      ] test_send_msg_invalid_agent
[       OK ] test_send_msg_invalid_agent
[ RUN      ] test_send_msg_disconnected_agent
[       OK ] test_send_msg_disconnected_agent
[ RUN      ] test_send_msg_encryption_error
[       OK ] test_send_msg_encryption_error
[ RUN      ] test_send_msg_tcp_ok
[       OK ] test_send_msg_tcp_ok
[ RUN      ] test_send_msg_tcp_err
[       OK ] test_send_msg_tcp_err
[ RUN      ] test_send_msg_tcp_err_closed_socket
[       OK ] test_send_msg_tcp_err_closed_socket
[ RUN      ] test_send_msg_udp_error
[       OK ] test_send_msg_udp_error
[ RUN      ] test_send_msg_udp_error_connection_reset
[       OK ] test_send_msg_udp_error_connection_reset
[ RUN      ] test_send_msg_udp_error_agent_not_responding
[       OK ] test_send_msg_udp_error_agent_not_responding
[ RUN      ] test_send_msg_udp_error_generic
[       OK ] test_send_msg_udp_error_generic
[==========] tests: 10 test(s) run.
[  PASSED  ] 10 test(s).
```

### CODE REVIEW

File: src/unit_tests/remoted/test_sendmsg.c [hunk 1]
```diff
@@ -39,32 +39,166 @@ static int group_teardown(void ** state) {
...
-static int test_setup_tcp(void ** state) {
+static int test_setup_keys(void ** state) {
     test_mode = 1;
 
     keyentry *node_key = NULL;
     os_calloc(1, sizeof(keyentry), node_key);
-    node_key->net_protocol = REMOTED_NET_PROTOCOL_TCP;
     node_key->rcvd = 10;
     node_key->sock = 15;
+    node_key->id = strdup("001");
```
Comment (MiguelazoDS) [inline src/unit_tests/remoted/test_sendmsg.c:49]: We usually use os_strdup() or w_strdup().

Comment (jftuduri) [reply src/unit_tests/remoted/test_sendmsg.c]: Right! Fixed.

File: src/unit_tests/remoted/test_sendmsg.c [hunk 2]
```diff
@@ -39,32 +39,166 @@ static int group_teardown(void ** state) {
...
+    assert_int_equal(ret, -1);
+    assert_int_equal(remoted_state.queued_msgs, 0);
+}
+
+void test_send_msg_encryption_error(void ** state) {
+    (void) state;
+
+    char *agent_id = "001";
+    char *msg = "abcdefghijk";
+    ssize_t msg_length = 11;
```
Comment (HanesSciarrone) [inline src/unit_tests/remoted/test_sendmsg.c:161]: Hi @jftuduri

Same change here

```suggestion
    ssize_t msg_length = strlen(msg);
```

File: src/unit_tests/remoted/test_sendmsg.c [hunk 3]
```diff
@@ -164,10 +298,313 @@ void test_send_msg_tcp_err(void ** state) {
     assert_int_equal(remoted_state.queued_msgs, 0);
 }
 
+void test_send_msg_tcp_err_closed_socket(void ** state) {
+    (void) state;
+
+    char *agent_id = "001";
+    char *msg = "abcdefghijk";
+    ssize_t msg_length = 11;
```
Comment (HanesSciarrone) [inline src/unit_tests/remoted/test_sendmsg.c:306]: Hi @jftuduri

Same change here

```suggestion
    ssize_t msg_length = strlen(msg);
```

File: src/unit_tests/remoted/test_sendmsg.c [hunk 4]
```diff
@@ -164,10 +298,313 @@ void test_send_msg_tcp_err(void ** state) {
...
+    int key = 0;
+
+    char *crypto_msg = "!@#123abc";
+    ssize_t crypto_size = 9;
```
Comment (HanesSciarrone) [inline src/unit_tests/remoted/test_sendmsg.c:310]: Hi @jftuduri

Same change here

```suggestion
    ssize_t crypto_size = strlen(msg);
```

File: src/unit_tests/remoted/test_sendmsg.c [hunk 5]
```diff
@@ -164,10 +298,313 @@ void test_send_msg_tcp_err(void ** state) {
...
+void test_send_msg_udp_ok(void ** state) {
+    (void) state;
+
+    char *agent_id = "001";
+    char *msg = "abcdefghijk";
+    ssize_t msg_length = 11;
+    int key = 0;
+
+    char *crypto_msg = "!@#123abc";
+    ssize_t crypto_size = 9;
```
Comment (HanesSciarrone) [inline src/unit_tests/remoted/test_sendmsg.c:357]: Hi @jftuduri

Same change here

```suggestion
    ssize_t crypto_size = strlen(msg);
```

File: src/unit_tests/remoted/test_sendmsg.c [hunk 6]
```diff
@@ -164,10 +298,313 @@ void test_send_msg_tcp_err(void ** state) {
...
+void test_send_msg_udp_error(void ** state) {
+    (void) state;
+
+    char *agent_id = "001";
+    char *msg = "abcdefghijk";
+    ssize_t msg_length = 11;
+    int key = 0;
+
+    char *crypto_msg = "!@#123abc";
+    ssize_t crypto_size = 9;
```
Comment (HanesSciarrone) [inline src/unit_tests/remoted/test_sendmsg.c:405]: Hi @jftuduri

Same change here

```suggestion
    ssize_t crypto_size = strlen(msg);
```

File: src/unit_tests/remoted/test_sendmsg.c [hunk 7]
```diff
@@ -164,10 +298,313 @@ void test_send_msg_tcp_err(void ** state) {
...
+    assert_int_equal(ret, -1);
+    assert_int_equal(remoted_state.queued_msgs, 0);
+}
+
+void test_send_msg_udp_error_connection_reset(void ** state) {
+    (void) state;
+
+    char *agent_id = "001";
+    char *msg = "abcdefghijk";
+    ssize_t msg_length = 11;
```
Comment (HanesSciarrone) [inline src/unit_tests/remoted/test_sendmsg.c:449]: Hi @jftuduri

Same change here

```suggestion
    ssize_t msg_length = strlen(msg);
```

File: src/unit_tests/remoted/test_sendmsg.c [hunk 8]
```diff
@@ -164,10 +298,313 @@ void test_send_msg_tcp_err(void ** state) {
...
+    int key = 0;
+
+    char *crypto_msg = "!@#123abc";
+    ssize_t crypto_size = 9;
```
Comment (HanesSciarrone) [inline src/unit_tests/remoted/test_sendmsg.c:453]: Hi @jftuduri

Same change here

```suggestion
    ssize_t crypto_size = strlen(msg);
```

File: src/unit_tests/remoted/test_sendmsg.c [hunk 9]
```diff
@@ -164,10 +298,313 @@ void test_send_msg_tcp_err(void ** state) {
...
+void test_send_msg_udp_error_agent_not_responding(void ** state) {
+    (void) state;
+
+    char *agent_id = "001";
+    char *msg = "abcdefghijk";
+    ssize_t msg_length = 11;
+    int key = 0;
+
+    char *crypto_msg = "!@#123abc";
+    ssize_t crypto_size = 9;
```
Comment (HanesSciarrone) [inline src/unit_tests/remoted/test_sendmsg.c:501]: Hi @jftuduri

Same change here

```suggestion
    ssize_t crypto_size = strlen(msg);
```

File: src/unit_tests/remoted/test_sendmsg.c [hunk 10]
```diff
@@ -164,10 +298,313 @@ void test_send_msg_tcp_err(void ** state) {
...
+    assert_int_equal(ret, -1);
+    assert_int_equal(remoted_state.queued_msgs, 0);
+}
+
+void test_send_msg_udp_error_agent_not_responding(void ** state) {
+    (void) state;
+
+    char *agent_id = "001";
+    char *msg = "abcdefghijk";
+    ssize_t msg_length = 11;
```
Comment (HanesSciarrone) [inline src/unit_tests/remoted/test_sendmsg.c:497]: Hi @jftuduri

Same change here

```suggestion
    ssize_t msg_length = strlen(msg);
```

File: src/unit_tests/remoted/test_sendmsg.c [hunk 11]
```diff
@@ -164,10 +298,313 @@ void test_send_msg_tcp_err(void ** state) {
...
+    assert_int_equal(ret, -1);
+    assert_int_equal(remoted_state.queued_msgs, 0);
+}
+
+void test_send_msg_udp_error_generic(void ** state) {
+    (void) state;
+
+    char *agent_id = "001";
+    char *msg = "abcdefghijk";
+    ssize_t msg_length = 11;
```
Comment (HanesSciarrone) [inline src/unit_tests/remoted/test_sendmsg.c:545]: Hi @jftuduri

Same change here

```suggestion
    ssize_t msg_length = strlen(msg);
```

File: src/unit_tests/remoted/test_sendmsg.c [hunk 12]
```diff
@@ -164,10 +298,313 @@ void test_send_msg_tcp_err(void ** state) {
...
+    int key = 0;
+
+    char *crypto_msg = "!@#123abc";
+    ssize_t crypto_size = 9;
```
Comment (HanesSciarrone) [inline src/unit_tests/remoted/test_sendmsg.c:549]: Hi @jftuduri

Same change here

```suggestion
    ssize_t crypto_size = strlen(msg);
```

**Review by HanesSciarrone — CHANGES_REQUESTED** [review]

**Review by pereyra-m — COMMENTED** [review]
GJ!

File: src/unit_tests/remoted/test_sendmsg.c (hunk 4 above)
Comment (jftuduri) [reply src/unit_tests/remoted/test_sendmsg.c]: Fixed

File: src/unit_tests/remoted/test_sendmsg.c [hunk 13]
```diff
@@ -164,10 +298,313 @@ void test_send_msg_tcp_err(void ** state) {
...
+    assert_int_equal(ret, -1);
+    assert_int_equal(remoted_state.queued_msgs, 0);
+}
+
+void test_send_msg_udp_ok(void ** state) {
+    (void) state;
+
+    char *agent_id = "001";
+    char *msg = "abcdefghijk";
+    ssize_t msg_length = 11;
```
[the remark this answers is not in the document]
Comment (jftuduri) [reply src/unit_tests/remoted/test_sendmsg.c]: Fixed

File: src/unit_tests/remoted/test_sendmsg.c (hunk 6 above)
Comment (jftuduri) [reply src/unit_tests/remoted/test_sendmsg.c]: Fixed

File: src/unit_tests/remoted/test_sendmsg.c (hunk 7 above)
Comment (jftuduri) [reply src/unit_tests/remoted/test_sendmsg.c]: Fixed

File: src/unit_tests/wrappers/linux/socket_wrappers.c [hunk 14]
```diff
@@ -57,6 +57,11 @@ ssize_t __wrap_send(__attribute__((unused))int __fd, __attribute__((unused))cons
     return mock();
 }
 
+ssize_t __wrap_sendto(__attribute__((unused)) int __fd, __attribute__((unused)) const void *__buf, __attribute__((unused)) size_t __n, __attribute__((unused)) int __flags,
+                      __attribute__((unused)) __CONST_SOCKADDR_ARG __addr, socklen_t __len){
```
[the remark this answers is not in the document]
Comment (jftuduri) [reply src/unit_tests/wrappers/linux/socket_wrappers.c]: Good catch! Fixed.

File: src/unit_tests/remoted/test_sendmsg.c [hunk 15]
```diff
@@ -164,10 +298,313 @@ void test_send_msg_tcp_err(void ** state) {
...
+    expect_value(__wrap_CreateSecMSG, msg_length, msg_length);
+    expect_value(__wrap_CreateSecMSG, id, key);
+    will_return(__wrap_CreateSecMSG, crypto_size);
+    will_return(__wrap_CreateSecMSG, crypto_msg);
+
+    expect_function_call(__wrap_pthread_mutex_lock);
+
+    will_return(__wrap_sendto, crypto_size);
+
+    //Shouldn't rem_add_send expect crypto_size instead of 0?
```
[the remark this answers is not in the document]
Comment (jftuduri) [reply src/unit_tests/remoted/test_sendmsg.c]: I removed the udp_ok UT until this bug is fixed. I will create an issue to fix the bug.

[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/unit_tests/remoted/test_sendmsg.c]: Great!
Don't forget to link/mention this issue in the new one!

**Review by pereyra-m — DISMISSED** [review]
LGTM!

**Review by HanesSciarrone — APPROVED** [review]

**Review by MiguelazoDS — DISMISSED** [review]

**Review by jnasselle — COMMENTED** [review]
Really nice work @jftuduri ! 
Could you please check const correctness of the new tests? is just a detail but part of coding style (and the proper way to do it)

**Review by Dwordcito — APPROVED** [review]

**Review by HanesSciarrone — APPROVED** [review]

**Review by jnasselle — APPROVED** [review]

**Review by MiguelazoDS — APPROVED** [review]

**Review by pereyra-m — APPROVED** [review]
LGTM!

**Pull request merged**